### ISSUE TITLE
64-bit NVDA: cannot read content from 32-bit SysListView32 implementations

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#18706]
Hi,

Part of #18207:

### Introduction
NVDA can read column header and content from 32-bit SysListView32 structures vis LVITEM structure definitions. However, this does not work on 64-bit NVDA because of differences in definitions between 32-bit and 64-bit SysListView32 structures. Any 32-bit SysListView32 controls are affected, among which is an interesting app called StationPlaylist Studio.

### Steps to reproduce:

Prerequisite: 64-bit NVDA must be prepared locally.

1. Run a 32-bit app with 32-bit SysListiew32 implementation.
2. Try using table navigation commands to read column header and content.

### Actual behavior:
NVDA stays silent.

### Expected behavior:
NVDA reports column header and content.

### NVDA logs, crash dumps and other attachments:
N/A

#### NVDA installed/portable/running from source:
Source

#### NVDA version:
alpha-51900,08399041 32-bit, wip-buildX64

#### Windows version:
Windows 11 24H2 beta (build 26120.5742)

#### Name and version of other software in use when reproducing the issue:
StationPlaylist Studio 6.11

#### Other information about your system:
Used as a development workstation, StationPlaylist is installed to work on the add-on

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No issues with 32-bit NVDA

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

#### 32-bit SysListiew32 LVITEM structure:
(source: NVDA source/NVDAObjects/IAccessible/syslistview32.py)

`
class LVITEM(Structure):  # noqa: F405
	_fields_ = [
		("mask", c_uint),  # noqa: F405
		("iItem", c_int),  # noqa: F405
		("iSubItem", c_int),  # noqa: F405
		("state", c_uint),  # noqa: F405
		("stateMask", c_uint),  # noqa: F405
		("pszText", c_void_p),  # noqa: F405
		("cchTextMax", c_int),  # noqa: F405
		("iImage", c_int),  # noqa: F405
		("lParam", LPARAM),  # noqa: F405
		("iIndent", c_int),  # noqa: F405
		("iGroupID", c_int),  # noqa: F405
		("cColumns", c_uint),  # noqa: F405
		("puColumns", c_uint),  # noqa: F405
		("piColFmt", POINTER(c_int)),  # noqa: F405
		("iGroup", c_int),  # noqa: F405
	]
`

#### 64-bit SysListView32 LVITEM structure:

`
class LVITEM64(Structure):  # noqa: F405
	_fields_ = [
		("mask", c_uint),  # noqa: F405
		("iItem", c_int),  # noqa: F405
		("iSubItem", c_int),  # noqa: F405
		("state", c_uint),  # noqa: F405
		("stateMask", c_uint),  # noqa: F405
		("pszText", c_ulonglong),  # noqa: F405
		("cchTextMax", c_int),  # noqa: F405
		("iImage", c_int),  # noqa: F405
		("lParam", c_ulonglong),  # noqa: F405
		("iIndent", c_int),  # noqa: F405
		("iGroupID", c_int),  # noqa: F405
		("cColumns", c_uint),  # noqa: F405
		("puColumns", c_uint),  # noqa: F405
		("piColFmt", c_ulonglong),  # noqa: F405
		("iGroup", c_int),  # noqa: F405
	]
`

#### 32-bit LVCOLUMN structure:

`
class LVCOLUMN(Structure):  # noqa: F405
	_fields_ = [
		("mask", c_uint),  # noqa: F405
		("fmt", c_int),  # noqa: F405
		("cx", c_int),  # noqa: F405
		("pszText", c_void_p),  # noqa: F405
		("cchTextMax", c_int),  # noqa: F405
		("iSubItem", c_int),  # noqa: F405
		("iImage", c_int),  # noqa: F405
		("iOrder", c_int),  # noqa: F405
		("cxMin", c_int),  # noqa: F405
		("cxDefault", c_int),  # noqa: F405
		("cxIdeal", c_int),  # noqa: F405
	]
`

#### 64-bit LVCOLUMN structure:

`
class LVCOLUMN64(Structure):  # noqa: F405
	_fields_ = [
		("mask", c_uint),  # noqa: F405
		("fmt", c_int),  # noqa: F405
		("cx", c_int),  # noqa: F405
		("pszText", c_ulonglong),  # noqa: F405
		("cchTextMax", c_int),  # noqa: F405
		("iSubItem", c_int),  # noqa: F405
		("iImage", c_int),  # noqa: F405
		("iOrder", c_int),  # noqa: F405
		("cxMin", c_int),  # noqa: F405
		("cxDefault", c_int),  # noqa: F405
		("cxIdeal", c_int),  # noqa: F405
	]
`

### Impact:
Table navigation on 32-bit SysListView32 will be broken with 64-bit transition. For people relying on 32-bit apps such as radio broadcasters using StationPlaylist Studio, this can lead to NVDA appearing to do nothing when reviewing list items.

Thanks.

**Comment by LeonarddeR** [comment]
Do you have any idea about the cause of this? Does `appModule.is64BitProcess` return the correct result?

**Comment by josephsl** [comment]
Hi, this is caused by dword values that should have been coded as integers (NVDA code). Thanks.

**Comment by josephsl** [comment]
Hi,

Update: issue was traced to NVDA's inability to communicate with 32-bit SysListView32 controls when (re)started while focused on 32-bit SysListView32 controls, with the specific issue being LVItem and LVColumn structure member definition (for 32-bit structures, void pointer (c_void_p) is used instead of unsigned long/int (remember that Windows uses "long long" for 64-bit)). This, together with removal of "import everything method" (from mod import *) will be part of an upcoming pull request.

Thanks.
